graftm/taxonomy.py

```diff
@@ -0,0 +1,176 @@
+#!/usr/bin/env python
+###############################################################################
+#                                                                             #
+# This program is free software: you can redistribute it and/or modify        #
+# it under the terms of the GNU General Public License as published by        #
+# the Free Software Foundation, either version 3 of the License, or           #
+# (at your option) any later version.                                         #
+#                                                                             #
+# This program is distributed in the hope that it will be useful,             #
+# but WITHOUT ANY WARRANTY; without even the implied warranty of              #
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                #
+# GNU General Public License for more details.                                #
+#                                                                             #
+# You should have received a copy of the GNU General Public License           #
+# along with this program. If not, see <http://www.gnu.org/licenses/>.        #
+#                                                                             #
+###############################################################################
+__author__ = "Joel Boyd, Ben Woodcroft"
+__copyright__ = "Copyright 2015"
+__credits__ = ["Joel Boyd"]
+__license__ = "GPL3"
+__version__ = "0.0.1"
+__maintainer__ = "Joel Boyd"
+__email__ = "joel.boyd near uq.net.au"
+__status__ = "Development"
+###############################################################################
+# System imports
+import logging 
+
+###############################################################################
+###############################################################################
+###############################################################################
+
+class MalformedGreenGenesFormattedFile(Exception):
+    pass
+
+#-------------------------------------------------------------------------------
+
+class ProgrammingError(Exception):
+    pass
+
+#-------------------------------------------------------------------------------
+
+class MissingTaxonomy(Exception):
+    pass
+
+#-------------------------------------------------------------------------------
+
+class Taxonomy:
+
+    TAB='\t'
+    COM=','
+    GG_SEP_0='; '
+    GG_SEP_1=';'
+    
+    def __init__(self):
+        self.taxonomy = {}
+        
+    def _ammend_taxonomy_gaps(self):
```

Comment:
So that when we search for the phylogenetic rank of a specific tip (as you mentioned above) the taxonomy[tip_name][rank] will not choke.